File: tutorial/fortran/04_privileges_accessor/privileges.f90
module daxpy
  use legion_fortran
  use iso_c_binding
  implicit none
  
  integer(c_int), parameter :: TOP_LEVEL_TASK_ID=0
  integer(c_int), parameter :: INIT_TASK_ID=1
  integer(c_int), parameter :: DAXPY_TASK_ID=2
  integer(c_int), parameter :: CHECK_TASK_ID=3
  
contains
  subroutine init_task(tdata, tdatalen, userdata, userlen, p)
    implicit none
    
    type(c_ptr), value, intent(in) :: tdata
    integer(c_size_t), value, intent(in) :: tdatalen
    type(c_ptr), value, intent(in) ::userdata
    integer(c_size_t), value, intent(in) :: userlen
    integer(c_long_long), value, intent(in) :: p
    
    type(FFieldAccessor1D) :: accessor
        
    type(legion_rect_1d_f_t) :: index_rect
    real(kind=8), target :: x_value
    type(FPoint1D) :: point_1d
    integer :: fid
    integer(kind=8) :: i
    
    type(FContext) :: ctx
    type(FRuntime) :: runtime
    type(FTask) :: task
    type(FPhysicalRegionList) :: pr_list
    type(FPhysicalRegion) :: physical_region
    type(FLogicalRegion) :: logical_region
    type(FRegionRequirement) :: region_requirement
    type(FIndexSpace) :: index_space
    type(FRect1D) :: rect_1d
      
    call legion_task_prolog(tdata, tdatalen, userdata, userlen, p, &
                            task, pr_list, &
                            ctx, runtime)
    
    region_requirement = task%get_region_requirement(0)
    fid = region_requirement%get_privilege_field(0)
                                
    physical_region = pr_list%get_region(0)                         
    
    accessor = FFieldAccessor1D(physical_region, fid, WRITE_DISCARD, c_sizeof(x_value))
    logical_region = region_requirement%get_region()
    index_space = logical_region%get_index_space()
    rect_1d = runtime%get_index_space_domain(ctx, index_space)
    index_rect = rect_1d%rect
    
    Print *, "Init Task!", fid
    
    do i = index_rect%lo%x(0), index_rect%hi%x(0)
        point_1d = FPoint1D(i)
        x_value = 1.1 * (real(fid,8)+1) + real(i,8)
        call accessor%write_point(point_1d, x_value)
    end do
    
    call legion_task_epilog(runtime, ctx)
  end subroutine init_task

  subroutine daxpy_task(tdata, tdatalen, userdata, userlen, p)
    implicit none
    
    type(c_ptr), value, intent(in) :: tdata
    integer(c_size_t), value, intent(in) :: tdatalen
    type(c_ptr), value, intent(in) ::userdata
    integer(c_size_t), value, intent(in) :: userlen
    integer(c_long_long), value, intent(in) :: p

    type(legion_rect_1d_f_t) :: index_rect
    real(kind=8), target :: z_value, x_value, y_value
    integer(kind=8) :: i
    integer(c_size_t) :: arglen
    real(kind=8), pointer :: task_arg
    type(c_ptr) :: task_arg_ptr
    real(kind=8) :: alpha
    
    type(FPoint1D) :: point_1d
    type(FFieldAccessor1D) :: accessor_x, accessor_y, accessor_z
    
    type(FContext) :: ctx
    type(FRuntime) :: runtime
    type(FTask) :: task
    type(FPhysicalRegionList) :: pr_list
    type(FPhysicalRegion) :: physical_region_0, physical_region_1
    type(FLogicalRegion) :: logical_region
    type(FRegionRequirement) :: region_requirement
    type(FIndexSpace) :: index_space
    type(FRect1D) :: rect_1d
      
    call legion_task_prolog(tdata, tdatalen, userdata, userlen, p, &
                            task, pr_list, &
                            ctx, runtime)
                            
    arglen = task%get_arglen()
    task_arg_ptr = task%get_args()
    call c_f_pointer(task_arg_ptr, task_arg)
    alpha = task_arg
    
    region_requirement = task%get_region_requirement(0)
                                
    physical_region_0 = pr_list%get_region(0)     
    physical_region_1 = pr_list%get_region(1)  
    print *, "daxpy num_regions", pr_list%num_regions      
    
    accessor_x = FFieldAccessor1D(physical_region_0, 0, READ_ONLY, c_sizeof(x_value))
    accessor_y = FFieldAccessor1D(physical_region_0, 1, READ_ONLY, c_sizeof(y_value))
    accessor_z = FFieldAccessor1D(physical_region_1, 2, WRITE_DISCARD, c_sizeof(z_value))
    logical_region = region_requirement%get_region()
    index_space = logical_region%get_index_space()
    rect_1d = runtime%get_index_space_domain(ctx, index_space)
    index_rect = rect_1d%rect
    
    Print *, "Daxpy Task!", alpha
    
    do i = index_rect%lo%x(0), index_rect%hi%x(0)
      point_1d = FPoint1D(i)
      call accessor_x%read_point(point_1d, x_value)
      call accessor_y%read_point(point_1d, y_value)
      z_value = alpha*x_value + y_value
      call accessor_z%write_point(point_1d, z_value)
    end do
    
    call legion_task_epilog(runtime, ctx)
  end subroutine daxpy_task

  subroutine check_task(tdata, tdatalen, userdata, userlen, p)
    implicit none
    
    type(c_ptr), value, intent(in) :: tdata
    integer(c_size_t), value, intent(in) :: tdatalen
    type(c_ptr), value, intent(in) ::userdata
    integer(c_size_t), value, intent(in) :: userlen
    integer(c_long_long), value, intent(in) :: p

    type(legion_rect_1d_f_t) :: index_rect
    real(kind=8), target :: x_value = 0
    real(kind=8), target :: y_value = 0
    real(kind=8), target :: z_value = 0
    integer(kind=8) :: i
    logical :: all_passed = .true.
    integer(c_size_t) :: arglen
    real(kind=8), pointer :: task_arg
    type(c_ptr) :: task_arg_ptr
    real(kind=8) :: alpha
    
    type(FPoint1D) :: point_1d
    type(FFieldAccessor1D) :: accessor_x, accessor_y, accessor_z
    
    type(FContext) :: ctx
    type(FRuntime) :: runtime
    type(FTask) :: task
    type(FPhysicalRegionList) :: pr_list
    type(FPhysicalRegion) :: physical_region_0, physical_region_1
    type(FLogicalRegion) :: logical_region
    type(FRegionRequirement) :: region_requirement
    type(FIndexSpace) :: index_space
    type(FRect1D) :: rect_1d
      
    integer :: count
      
    call legion_task_prolog(tdata, tdatalen, userdata, userlen, p, &
                            task, pr_list, &
                            ctx, runtime)
    
    arglen = task%get_arglen()
    task_arg_ptr = task%get_args()
    call c_f_pointer(task_arg_ptr, task_arg)
    alpha = task_arg
    
    region_requirement = task%get_region_requirement(0)
                                
    physical_region_0 = pr_list%get_region(0)     
    physical_region_1 = pr_list%get_region(1)        
    
    accessor_x = FFieldAccessor1D(physical_region_0, 0, READ_ONLY, c_sizeof(x_value))
    accessor_y = FFieldAccessor1D(physical_region_0, 1, READ_ONLY, c_sizeof(y_value))
    accessor_z = FFieldAccessor1D(physical_region_1, 2, READ_ONLY, c_sizeof(z_value))
    logical_region = region_requirement%get_region()
    index_space = logical_region%get_index_space()
    rect_1d = runtime%get_index_space_domain(ctx, index_space)
    index_rect = rect_1d%rect
    
    count = 0
    do i = index_rect%lo%x(0), index_rect%hi%x(0)
      count = count + 1
      point_1d = FPoint1D(i)
      call accessor_x%read_point(point_1d, x_value)
      call accessor_y%read_point(point_1d, y_value)
      call accessor_z%read_point(point_1d, z_value)
      if (alpha*x_value + y_value == z_value) then
      else
        print *, "wrong", i, x_value, y_value, z_value
        all_passed = .false.
      end if
    end do
    
    if (all_passed .eqv. .true.) then
      print *, "Pass", count
    else
      print *, "Failed"
    end if
    
    call legion_task_epilog(runtime, ctx)
  end subroutine check_task

  subroutine top_level_task(tdata, tdatalen, userdata, userlen, p)
    implicit none
    
    type(c_ptr), value, intent(in) :: tdata
    integer(c_size_t), value, intent(in) :: tdatalen
    type(c_ptr), value, intent(in) ::userdata
    integer(c_size_t), value, intent(in) :: userlen
    integer(c_long_long), value, intent(in) :: p
    
    real(kind=8) :: real_number = 0.0
    real(kind=8), target :: alpha = 0.1    

    integer*4 :: num_elements = 1024
    
    type(FRect1D) :: elem_rect
    type(FRuntime) :: runtime
    type(FContext) :: ctx
    type(FTask) :: task
    type(FPhysicalRegionList) :: pr_list
    type(FIndexSpace) :: is
    type(FFieldSpace) :: input_fs, output_fs
    type(FFieldAllocator) :: ifs_allocator, ofs_allocator
    type(FLogicalRegion) :: input_lr, output_lr
    type(FTaskLauncher) :: init_launcher_x, init_launcher_y, daxpy_launcher, check_launcher
    type(FFuture) ::task_future
    
    Print *, "TOP Level Task!"
    
    call legion_task_prolog(tdata, tdatalen, userdata, userlen, p, &
                            task, pr_list, &
                            ctx, runtime)
    
    ! create index space, field space and logical region
    elem_rect = FRect1D(0, num_elements-1)
    is = runtime%create_index_space(ctx, elem_rect)
    
    input_fs = runtime%create_field_space(ctx)
    ifs_allocator = runtime%create_field_allocator(ctx, input_fs)
    call ifs_allocator%allocate_field(c_sizeof(real_number), 0)
    call ifs_allocator%allocate_field(c_sizeof(real_number), 1)
    call ifs_allocator%destroy()
    
    output_fs = runtime%create_field_space(ctx)
    ofs_allocator = runtime%create_field_allocator(ctx, output_fs)
    call ofs_allocator%allocate_field(c_sizeof(real_number), 2)
    call ofs_allocator%destroy()
    
    input_lr = runtime%create_logical_region(ctx, is, input_fs)
    output_lr = runtime%create_logical_region(ctx, is, output_fs)
    
    !init task for X
    init_launcher_x = FTaskLauncher(INIT_TASK_ID, FTaskArgument())
    call init_launcher_x%add_region_requirement(input_lr, & 
                                              WRITE_DISCARD, EXCLUSIVE, &
                                              input_lr)                                          
    call init_launcher_x%add_field(0, 0)
    task_future = runtime%execute_task(ctx, init_launcher_x)
    
    !init task for Y
    init_launcher_y = FTaskLauncher(INIT_TASK_ID, FTaskArgument())
    call init_launcher_y%add_region_requirement(input_lr, & 
                                              WRITE_DISCARD, EXCLUSIVE, &
                                              input_lr)                                          
    call init_launcher_y%add_field(0, 1)
    task_future = runtime%execute_task(ctx, init_launcher_y)
    
    !daxpy task
    daxpy_launcher = FTaskLauncher(DAXPY_TASK_ID, FTaskArgument(c_loc(alpha), c_sizeof(alpha)))
    call daxpy_launcher%add_region_requirement(input_lr, & 
                                              READ_ONLY, EXCLUSIVE, &
                                              input_lr)                                          
    call daxpy_launcher%add_field(0, 0)
    call daxpy_launcher%add_field(0, 1)
    call daxpy_launcher%add_region_requirement(output_lr, & 
                                              WRITE_DISCARD, EXCLUSIVE, &
                                              output_lr)                                          
    call daxpy_launcher%add_field(1, 2)
    task_future = runtime%execute_task(ctx, daxpy_launcher)
    
    !check task
    check_launcher = FTaskLauncher(CHECK_TASK_ID, FTaskArgument(c_loc(alpha), c_sizeof(alpha)))
    call check_launcher%add_region_requirement(input_lr, & 
                                              READ_ONLY, EXCLUSIVE, &
                                              input_lr)                                          
    call check_launcher%add_field(0, 0)
    call check_launcher%add_field(0, 1)
    call check_launcher%add_region_requirement(output_lr, & 
                                              READ_ONLY, EXCLUSIVE, &
                                              output_lr)                                          
    call check_launcher%add_field(1, 2)
    task_future = runtime%execute_task(ctx, check_launcher) 
    
    ! clean up
    call runtime%destroy_logical_region(ctx, input_lr)
    call runtime%destroy_logical_region(ctx, output_lr)
    call runtime%destroy_field_space(ctx, input_fs)
    call runtime%destroy_field_space(ctx, output_fs)
    call runtime%destroy_index_space(ctx, is)
    call init_launcher_x%destroy()
    call init_launcher_y%destroy()
    call daxpy_launcher%destroy()
    call check_launcher%destroy()
    
    call legion_task_epilog(runtime, ctx)
  end subroutine top_level_task
end module daxpy

Program daxpy_1d_accessor
  use legion_fortran
  use iso_c_binding
  use daxpy
  implicit none
  
  type(FTaskVariantRegistrar) :: registrar_top, registrar_init, registrar_daxpy, registrar_check
      
  call set_top_level_task_id(TOP_LEVEL_TASK_ID)
  
  registrar_top = FTaskVariantRegistrar(TOP_LEVEL_TASK_ID)
  call registrar_top%add_constraint(FProcessorConstraint(LOC_PROC))
  call preregister_task_variant(top_level_task, registrar_top, "top_level_task")
  call registrar_top%destroy()
  
  registrar_init = FTaskVariantRegistrar(INIT_TASK_ID)
  call registrar_init%add_constraint(FProcessorConstraint(LOC_PROC))
  call registrar_init%set_leaf()
  call preregister_task_variant(init_task, registrar_init, "init_task")
  call registrar_init%destroy()
  
  registrar_daxpy = FTaskVariantRegistrar(DAXPY_TASK_ID)
  call registrar_daxpy%add_constraint(FProcessorConstraint(LOC_PROC))
  call registrar_daxpy%set_leaf()
  call preregister_task_variant(daxpy_task, registrar_daxpy, "daxpy_task")
  call registrar_daxpy%destroy()
  
  registrar_check = FTaskVariantRegistrar(CHECK_TASK_ID)
  call registrar_check%add_constraint(FProcessorConstraint(LOC_PROC))
  call registrar_check%set_leaf()
  call preregister_task_variant(check_task, registrar_check, "check_task")
  call registrar_check%destroy()
  
  call legion_runtime_start()
End Program daxpy_1d_accessor

File: tutorial/06_privileges/privileges.cc
/* Copyright 2022 Stanford University
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


#include <cstdio>
#include <cassert>
#include <cstdlib>
#include "legion.h"
using namespace Legion;

enum TaskIDs {
  TOP_LEVEL_TASK_ID,
  INIT_FIELD_TASK_ID,
  DAXPY_TASK_ID,
  CHECK_TASK_ID,
};

enum FieldIDs {
  FID_X,
  FID_Y,
  FID_Z,
};

void top_level_task(const Task *task,
                    const std::vector<PhysicalRegion> &regions,
                    Context ctx, Runtime *runtime)
{
  int num_elements = 1024; 
  // See if we have any command line arguments to parse
  {
    const InputArgs &command_args = Runtime::get_input_args();
    for (int i = 1; i < command_args.argc; i++)
    {
      if (!strcmp(command_args.argv[i],"-n"))
        num_elements = atoi(command_args.argv[++i]);
    }
  }
  printf("Running daxpy for %d elements...\n", num_elements);

  // Create our logical regions using the same schema that
  // we used in the previous example.
  const Rect<1> elem_rect(0,num_elements-1);
  IndexSpace is = runtime->create_index_space(ctx, elem_rect); 
  FieldSpace input_fs = runtime->create_field_space(ctx);
  {
    FieldAllocator allocator = 
      runtime->create_field_allocator(ctx, input_fs);
    allocator.allocate_field(sizeof(double),FID_X);
    allocator.allocate_field(sizeof(double),FID_Y);
  }
  FieldSpace output_fs = runtime->create_field_space(ctx);
  {
    FieldAllocator allocator = 
      runtime->create_field_allocator(ctx, output_fs);
    allocator.allocate_field(sizeof(double),FID_Z);
  }
  LogicalRegion input_lr = runtime->create_logical_region(ctx, is, input_fs);
  LogicalRegion output_lr = runtime->create_logical_region(ctx, is, output_fs);

  // Instead of using an inline mapping to initialize the fields for
  // daxpy, in this case we will launch two separate tasks for initializing
  // each of the fields in parallel.  To launch the sub-tasks for performing
  // the initialization we again use the launcher objects that were
  // introduced earlier.  The only difference now is that instead of passing
  // arguments by value, we now want to specify the logical regions
  // that the tasks may access as their arguments.  We again make use of
  // the RegionRequirement struct to name the logical regions and fields
  // for which the task should have privileges.  In this case we launch
  // a task that asks for WRITE_DISCARD privileges on the 'X' field.
  //
  // An important property of the Legion programming model is that sub-tasks
  // are only allowed to request privileges which are a subset of a 
  // parent task's privileges.  When a task creates a logical region it
  // is granted full read-write privileges for that logical region.  It
  // can then pass them down to sub-tasks.  In this example the top-level
  // task has full privileges on all the fields of input_lr and output_lr.
  // In this call it passing read-write privileges down to the sub-task
  // on input_lr on field 'X'.  Legion will enforce the property that the 
  // sub-task only accesses the 'X' field of input_lr.  This property of
  // Legion is crucial for the implementation of Legion's hierarchical
  // scheduling algorithm which is described in detail in our two papers.
  TaskLauncher init_launcher(INIT_FIELD_TASK_ID, TaskArgument(NULL, 0));
  init_launcher.add_region_requirement(
      RegionRequirement(input_lr, WRITE_DISCARD, EXCLUSIVE, input_lr));
  init_launcher.add_field(0/*idx*/, FID_X);
  // Note that when we launch this task we don't record the future.
  // This is because we're going to let Legion be responsible for 
  // computing the data dependences between how different tasks access
  // logical regions.
  runtime->execute_task(ctx, init_launcher);

  // Re-use the same launcher but with a slightly different RegionRequirement
  // that requests privileges on field 'Y' instead of 'X'.  Since these
  // two instances of the init_field_task are accessing different fields
  // of the input_lr region, they can be run in parallel (whether or not
  // they do is dependent on the mapping discussed in a later example).
  // Legion automatically will discover this parallelism since the runtime
  // understands the fields present on the logical region.
  //
  // We now call attention to a unique property of the init_field_task.
  // In this example we've actually called the task with two different
  // region requirements containing different fields.  The init_field_task
  // is an example of a field-polymorphic task which is capable of 
  // performing the same operation on different fields of a logical region.
  // In practice this is very useful property for a task to maintain as
  // it allows one implementation of a task to be written which is capable
  // of being used in many places.
  init_launcher.region_requirements[0].privilege_fields.clear();
  init_launcher.region_requirements[0].instance_fields.clear();
  init_launcher.add_field(0/*idx*/, FID_Y);

  runtime->execute_task(ctx, init_launcher);

  // Now we launch the task to perform the daxpy computation.  We pass
  // in the alpha value as an argument.  All the rest of the arguments
  // are RegionRequirements specifying that we are reading the two
  // fields on the input_lr region and writing results to the output_lr
  // region.  Legion will automatically compute data dependences
  // from the two init_field_tasks and will ensure that the program
  // order execution is obeyed.
  const double alpha = drand48();
  TaskLauncher daxpy_launcher(DAXPY_TASK_ID, TaskArgument(&alpha, sizeof(alpha)));
  daxpy_launcher.add_region_requirement(
      RegionRequirement(input_lr, READ_ONLY, EXCLUSIVE, input_lr));
  daxpy_launcher.add_field(0/*idx*/, FID_X);
  daxpy_launcher.add_field(0/*idx*/, FID_Y);
  daxpy_launcher.add_region_requirement(
      RegionRequirement(output_lr, WRITE_DISCARD, EXCLUSIVE, output_lr));
  daxpy_launcher.add_field(1/*idx*/, FID_Z);

  runtime->execute_task(ctx, daxpy_launcher);

  // Finally we launch a task to perform the check on the output.  Note
  // that Legion will compute a data dependence on the first RegionRequirement
  // with the two init_field_tasks, but not on daxpy task since they 
  // both request read-only privileges on the 'X' and 'Y' fields.  However,
  // Legion will compute a data dependence on the second region requirement
  // as the daxpy task was writing the 'Z' field on output_lr and this task
  // is reading the 'Z' field of the output_lr region.
  TaskLauncher check_launcher(CHECK_TASK_ID, TaskArgument(&alpha, sizeof(alpha)));
  check_launcher.add_region_requirement(
      RegionRequirement(input_lr, READ_ONLY, EXCLUSIVE, input_lr));
  check_launcher.add_field(0/*idx*/, FID_X);
  check_launcher.add_field(0/*idx*/, FID_Y);
  check_launcher.add_region_requirement(
      RegionRequirement(output_lr, READ_ONLY, EXCLUSIVE, output_lr));
  check_launcher.add_field(1/*idx*/, FID_Z);

  runtime->execute_task(ctx, check_launcher);

  // Notice that we never once blocked waiting on the result of any sub-task
  // in the execution of the top-level task.  We don't even block before
  // destroying any of our resources.  This works because Legion understands
  // the data being accessed by all of these operations and defers all of
  // their executions until they are safe to perform.  Legion is still smart
  // enough to know that the top-level task is not finished until all of
  // the sub operations that have been performed are completed.  However,
  // from the programmer's perspective, all of these operations can be
  // done without ever blocking and thereby exposing as much task-level
  // parallelism to the Legion runtime as possible.  We'll discuss the
  // implications of Legion's deferred execution model in a later example.
  runtime->destroy_logical_region(ctx, input_lr);
  runtime->destroy_logical_region(ctx, output_lr);
  runtime->destroy_field_space(ctx, input_fs);
  runtime->destroy_field_space(ctx, output_fs);
  runtime->destroy_index_space(ctx, is);
}

// Note that tasks get a physical region for every region requirement
// that they requested when they were launched in the vector of 'regions'.
// In some cases the mapper may have chosen not to map the logical region
// which means that the task has the necessary privileges to access the
// region but not a physical instance to access.
void init_field_task(const Task *task,
                     const std::vector<PhysicalRegion> &regions,
                     Context ctx, Runtime *runtime)
{
  // Check that the inputs look right since we have no
  // static checking to help us out.
  assert(regions.size() == 1); 
  assert(task->regions.size() == 1);
  assert(task->regions[0].privilege_fields.size() == 1);
  // This is a field polymorphic function so figure out
  // which field we are responsible for initializing.
  FieldID fid = *(task->regions[0].privilege_fields.begin());
  printf("Initializing field %d...\n", fid);
  // Note that Legion's default mapper always map regions
  // and the Legion runtime is smart enough not to start
  // the task until all the regions contain valid data.  
  // Therefore in this case we don't need to call 'wait_until_valid'
  // on our physical regions and we know that getting this
  // accessor will never block the task's execution.  If
  // however we chose to unmap this physical region and then
  // remap it then we would need to call 'wait_until_valid'
  // again to ensure that we were accessing valid data.
  const FieldAccessor<WRITE_DISCARD,double,1> acc(regions[0], fid);

  Rect<1> rect = runtime->get_index_space_domain(ctx, 
                  task->regions[0].region.get_index_space());
  for (PointInRectIterator<1> pir(rect); pir(); pir++)
    acc[*pir] = drand48();
}

void daxpy_task(const Task *task,
                const std::vector<PhysicalRegion> &regions,
                Context ctx, Runtime *runtime)
{
  assert(regions.size() == 2);
  assert(task->regions.size() == 2);
  assert(task->arglen == sizeof(double));
  const double alpha = *((const double*)task->args);

  const FieldAccessor<READ_ONLY,double,1> acc_x(regions[0], FID_X);
  const FieldAccessor<READ_ONLY,double,1> acc_y(regions[0], FID_Y);
  const FieldAccessor<WRITE_DISCARD,double,1> acc_z(regions[1], FID_Z);

  printf("Running daxpy computation with alpha %.8g...\n", alpha);
  Rect<1> rect = runtime->get_index_space_domain(ctx,
                  task->regions[0].region.get_index_space());
  for (PointInRectIterator<1> pir(rect); pir(); pir++)
    acc_z[*pir] = alpha * acc_x[*pir] + acc_y[*pir];
}

void check_task(const Task *task,
                const std::vector<PhysicalRegion> &regions,
                Context ctx, Runtime *runtime)
{
  assert(regions.size() == 2);
  assert(task->regions.size() == 2);
  assert(task->arglen == sizeof(double));
  const double alpha = *((const double*)task->args);
  const FieldAccessor<READ_ONLY,double,1> acc_x(regions[0], FID_X);
  const FieldAccessor<READ_ONLY,double,1> acc_y(regions[0], FID_Y);
  const FieldAccessor<READ_ONLY,double,1> acc_z(regions[1], FID_Z);

  printf("Checking results...");
  Rect<1> rect = runtime->get_index_space_domain(ctx,
                  task->regions[0].region.get_index_space());
  bool all_passed = true;
  for (PointInRectIterator<1> pir(rect); pir(); pir++)
  {
    double expected = alpha * acc_x[*pir] + acc_y[*pir];
    double received = acc_z[*pir];
    // Probably shouldn't check for floating point equivalence but
    // the order of operations are the same should they should
    // be bitwise equal.
    if (expected != received) {
      all_passed = false;
      printf("expected %f, received %f\n", expected, received);
    }
  }
  if (all_passed)
    printf("SUCCESS!\n");
  else {
    printf("FAILURE!\n");
    abort();
  }
}

int main(int argc, char **argv)
{
  Runtime::set_top_level_task_id(TOP_LEVEL_TASK_ID);

  {
    TaskVariantRegistrar registrar(TOP_LEVEL_TASK_ID, "top_level");
    registrar.add_constraint(ProcessorConstraint(Processor::LOC_PROC));
    Runtime::preregister_task_variant<top_level_task>(registrar, "top_level");
  }

  {
    TaskVariantRegistrar registrar(INIT_FIELD_TASK_ID, "init_field");
    registrar.add_constraint(ProcessorConstraint(Processor::LOC_PROC));
    registrar.set_leaf();
    Runtime::preregister_task_variant<init_field_task>(registrar, "init_field");
  }

  {
    TaskVariantRegistrar registrar(DAXPY_TASK_ID, "daxpy");
    registrar.add_constraint(ProcessorConstraint(Processor::LOC_PROC));
    registrar.set_leaf();
    Runtime::preregister_task_variant<daxpy_task>(registrar, "daxpy");
  }

  {
    TaskVariantRegistrar registrar(CHECK_TASK_ID, "check");
    registrar.add_constraint(ProcessorConstraint(Processor::LOC_PROC));
    registrar.set_leaf();
    Runtime::preregister_task_variant<check_task>(registrar, "check");
  }

  return Runtime::start(argc, argv);
}
